### PR TITLE
Fix #283: Change bool functions to m bool

### LINKED ISSUE #283: `isWorthRetrying` (and other `A -> Bool` functions) should allow for effects (`A -> M[Bool]`)
This was we can retry on effects. The current behavior would be easily replicated by calling `pure` on the returned Bool. It is also what the haskell retry library does. 

Would require a change in the interface. Migration should be simple though.

### PR DESCRIPTION
**Pull request by pjrt**
This is what the retry haskell library does, and allows to retry on effectful operations.

### CODE REVIEW

File: modules/core/shared/src/test/scala/retry/PackageObjectSpec.scala [hunk 1]
```diff
@@ -398,7 +398,7 @@ class PackageObjectSpec extends AnyFlatSpec {
 
     val finalResult = retryingOnFailuresAndAllErrors[String](
       policy,
-      _ == "will never happen",
+      s => Right(s == "will never happen"),
       onError,
       onError
     ) {
```
Comment (cb372) [inline modules/core/shared/src/test/scala/retry/PackageObjectSpec.scala:404]: I think it's worth adding a couple of tests to show what happens if `wasSuccessful` or `isWorthRetrying` raises an error.

**Comment by cb372** [comment]
Hey, sorry about the delay in reviewing.

LGTM apart from my comment about adding a couple of tests.

I was wondering why the bincompat check didn't fail. Looks like we disabled the mima check in preparation for releasing v2.0, and forgot to re-enable them 😅

File: modules/core/shared/src/test/scala/retry/PackageObjectSpec.scala (hunk 1 above)
Comment (pjrt) [reply modules/core/shared/src/test/scala/retry/PackageObjectSpec.scala:423]: Done.

**Comment by pjrt** [comment]
@cb372 no prob. It sounds like we wanna hold this until cats 3, so no rush.

**Comment by wogan** [comment]
Maybe consider leaving the old signatures as overloads that use M.pure?

**Review by cb372 — APPROVED** [review]

**Comment by cb372** [comment]
> Maybe consider leaving the old signatures as overloads that use M.pure?

I can see the value in that, as it would ease migration and might be convenient for some use cases. But I'd like to keep the API as small as possible and in general I try to avoid overloads.

We could provide a scalafix rewrite and register it with scala steward?

**Pull request merged**